crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,483 @@
+/// @param gate The StandardGate to add to the circuit.
+/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on.
+/// @param params The pointer to the array of ``double`` values to use for the gate parameters.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///
+/// # Safety
+///
+/// The ``qubits`` and ``params`` types are expected to be a non-null pointer to an array of
+/// ``uint32_t`` and ``double`` respectively where the length is matching the expectations for
+/// the standard gate. If the array is insufficently long the behavior of this function is
+/// undefined as this will read outside the bounds of the array.
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_append_standard_gate(
+    circuit: *mut CircuitData,
+    gate: StandardGate,
+    qubits: *const u32,
+    params: *const f64,
+) -> ExitCode {
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+    unsafe {
+        let qargs: &[Qubit] = match gate.num_qubits() {
+            0 => &[],
+            1 => &[Qubit(*qubits.wrapping_add(0))],
+            2 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+            ],
+            3 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+                Qubit(*qubits.wrapping_add(2)),
+            ],
+            // There are no standard gates > 3 qubits
+            _ => unreachable!(),
+        };
+        let params: &[Param] = match gate.num_params() {
+            0 => &[],
+            1 => &[(*params.wrapping_add(0)).into()],
+            2 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+            ],
+            3 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+                (*params.wrapping_add(2)).into(),
+            ],
+            // There are no standard gates that take > 3 params
+            _ => unreachable!(),
+        };
+        circuit.push_standard_gate(gate, params, qargs);
+    }
+    ExitCode::Success
+}
+
+/// @ingroup QkCircuit
+/// Append a measurement to the circuit
+///
+/// @param circuit A pointer to the circuit to add the gate to
+/// @param qubits The ``uint32_t`` for the qubit to measure
+/// @param clbits The ``uint32_t`` for the clbit to store the measurement outcome in
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_measure(qc, 0, 0);
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_append_measure(
```

Comment:
Should we just call this `qk_circuit_measure` (same for the other directives)? The `append` doesn't really add any information, or what do you think?

Reply:
I can drop append if you prefer. I named it this way because of the gate function I used append.

Reply:
Adding to the bikeshedding, but generally I'm with Julien - the `append` is just noise (and we don't have it in Python).  Similarly, I'd suggest `qk_circuit_append_standard_gate` just becomes `qk_circuit_gate`, since `StandardGate` should be _by far_ the most common thing to add. Once we support custom instructions, we can call that something like `qk_circuit_custom`.

Reply:
Renamed in: https://github.com/Qiskit/qiskit/pull/14006/commits/1240467ab1fba3a9ae3c57a514d9f272c235fe56